01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+to them on stdin. I'm going to use the `Bits` type just to represent which calls
+have state on disk to either access (like `put`) or return (like `get`). This
+would normally just be the working directory of the script.
+
+## General Types
+
+```elm
+-- arbitrary configuration
+type alias Config = Dict String Json.Value
+
+-- identifier for space, i.e. {"branch":"foo"}
+type alias Space = Dict String String
+
+-- identifier for version, i.e. {"version":"1.2"}
+type alias Version = Dict String String
+
+-- arbitrary ordered metadata (we may make this fancier in the future)
+type alias Metadata = List (String, String)
+
+-- data on disk
+type alias Bits = ()
+```
+
+## Versioned Artifacts interface
+
+```elm
+spaces  : Config -> Set Space
+check   : Config -> Space -> Maybe Version -> List (Version, Metadata)
+get     : Config -> Space -> Version -> Bits
+put     : Config -> Bits -> (Space, Set Version)
+destroy : Config -> Bits -> Dict Space (Set Version)
+```
+
+## Notifications interface
+
+```elm
+notify : Config -> Notification -> ()
+
+type Notification
+  = BuildStarted
+      { build : BuildMetadata
+      , self : Maybe BuildInput
+      }
+  | BuildFinished
+      { build : BuildMetadata
+      , status : Status
+      , self : Maybe BuildInput
+      }
+
+type alias BuildMetadata =
+  { teamName : String
+  , pipelineName : String
+  , jobName : String
+  , buildName : String
+  , buildID : Int
+  , status : String
+  }
+
+type alias BuildInput =
+  { space : Space
+  , version : Version
+  }
+```
+
+
+# Examples
+
+TODO:
+
+- Pull Requests
+- Feature branches
+- Build matrixes
+- Generating branches (and propagating them downstream)
+- Semver artifacts
+- Fanning out against multiple IaaSes
+- Pool resource?
+- BOSH deploys
+
+
+# Summary of Changes
+
+## Overarching Changes
+
+* Add a `info` script which prints the resource's API version, e.g.
+  `{"version":"2.0"}`. This will start at `2.0`. If `/info` does not exist we'll
+  execute today's resource interface behavior.
+
+* Rather than running `/opt/resource/X`, discover the supported resource APIs by
+  invoking `info`. This allows us to be more flexible in what kinds of resources
+  we can support (versioned artifacts, notifications, ???), and where the
+  scripts may live (`/opt/resource` is very Linux specific).
+
+* Today's resource interface (`/in`, `/out`, `/check`) becomes more specifically
+  a "versioned artifacts" or just "artifacts" resource interface.
+
+* Introduction of some sort of schema validation for resource configuration.
+
+* Remove the distinction between `source` and `params`; resources will receive a
+  single `config`. The distinction will remain in the pipeline. This makes it
+  easier to implement a resource without planning ahead for interesting dynamic
+  vs. static usage patterns, and will get more powerful with #684.
+
+
+## Changes to Versioned Artifact resources
+
+* Add a `spaces` actopm, which is used to discover spaces.
```

Comment:
Spaces seem to be tied to stuff that can be discovered by resources (e.g. branches), but I'm not sure I see anything in the RFC in terms of build matrices - at least not the type that was in the original proposal for spaces, such as specifying the ruby version?

Also, would it be possible for a resource to support multiple `spaces`? For instance, the `git` and `github-pullrequest` resources already have `path` and `ignore_paths` parameters. Could `path` theoretically be a space on these resources as well?

Example:
```yml
resources:
- name: atc
  type: git
  source: {uri: https://github.com/concourse/atc}
  spaces:
    branch: [master]
    path: [concourse/atc/*, concourse/fly/* ]
```

So a commit to master which changes both paths would produce two versions, each with a different path in their `spaces` payload. The reason I'm asking about this is because it would allow us to have spaces for monolithic repositories. Would it would be considered an anti-pattern to do it this way, since what would be pulled in `get` would be the complete repository? Also, would this put too much responsibility on the resources in order to implement all the "spaces" required to make build matrices?

Reply:
Correct - spaces would only be useful for build matrixes when the things you're matrixing over can be represented as resources. Our theory/hope is that this should cover most things. For example, matrixing over Ruby versions would be done by using the Ruby docker images, which are semver'd:

```yaml
resources:
- name: ruby
  type: docker-image
  source: {repository: ruby}

jobs:
- name: unit
  plan:
  - get: ruby
    spaces: [{tag: 1.*}, {tag: 2.*}]
  - get: my-repo
  - task: unit
    file: my-repo/ci/unit.yml
    image: ruby
```

We don't currently have a plan for matrixes that are just static config (e.g. `params` to tasks). Spaces is a resource-only feature right now, and it may be unwise to stretch its scope any further at this point. I suspect this may end up leading to more manageable patterns anyway - by forcing everything into resources you'll be working with checked-in code/config rather than reconfiguring pipelines.

As long as it's easy to implement custom resource types, we could see specialized resource types for supporting workflows like giant monorepos and path-based spaces. I see the future tending towards many workflow-specific resource types vs. one big swiss-army `git` resource type.